nexus/db-queries/src/db/datastore/support_bundle.rs

```diff
@@ -0,0 +1,1298 @@
+
+use super::DataStore;
+use crate::authz;
+use crate::context::OpContext;
+use crate::db;
+use crate::db::error::public_error_from_diesel;
+use crate::db::error::ErrorHandler;
+use crate::db::model::Dataset;
+use crate::db::model::DatasetKind;
+use crate::db::model::SupportBundle;
+use crate::db::model::SupportBundleState;
+use crate::db::pagination::paginated;
+use crate::db::update_and_check::{UpdateAndCheck, UpdateStatus};
+use crate::transaction_retry::OptionalError;
+use async_bb8_diesel::AsyncRunQueryDsl;
+use diesel::prelude::*;
+use futures::FutureExt;
+use nexus_types::identity::Asset;
+use omicron_common::api::external;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::DataPageParams;
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupResult;
+use omicron_uuid_kinds::BlueprintUuid;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::OmicronZoneUuid;
+use omicron_uuid_kinds::SupportBundleUuid;
+use omicron_uuid_kinds::ZpoolUuid;
+use uuid::Uuid;
+
+const CANNOT_ALLOCATE_ERR_MSG: &'static str =
+"Current policy limits support bundle creation to 'one per external disk', and \
+ no disks are available. Either delete old support bundles or add additional \
+ disks";
+
+const FAILURE_REASON_NO_DATASET: &'static str =
+    "Allocated dataset no longer exists";
+const FAILURE_REASON_NO_NEXUS: &'static str =
+    "Nexus managing this bundle no longer exists";
+
+/// Provides a report on how many bundle were expunged, and why.
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct SupportBundleExpungementReport {
+    /// Bundles marked "failed" because the datasets storing them have been
+    /// expunged.
+    pub bundles_failed_missing_datasets: usize,
+    /// Bundles deleted because the datasets storing them have been
+    /// expunged.
+    pub bundles_deleted_missing_datasets: usize,
+
+    /// Bundles marked "destroying" because the nexuses managing them have been
+    /// expunged.
+    ///
+    /// These bundles should be re-assigned to a different nexus for cleanup.
+    pub bundles_failing_missing_nexus: usize,
+
+    /// Bundles which had a new Nexus assigned to them.
+    pub bundles_reassigned: usize,
+}
+
+impl DataStore {
+    /// Creates a new support bundle.
+    ///
+    /// Requires that the UUID of the calling Nexus be supplied as input -
+    /// this particular Zone is responsible for the collection process.
+    ///
+    /// Note that really any functioning Nexus would work as the "assignee",
+    /// but it's clear that our instance will work, because we're currently
+    /// running.
+    pub async fn support_bundle_create(
+        &self,
+        opctx: &OpContext,
+        reason_for_creation: &'static str,
+        this_nexus_id: OmicronZoneUuid,
+    ) -> CreateResult<SupportBundle> {
+        opctx.authorize(authz::Action::ListChildren, &authz::FLEET).await?;
```

Comment:
Since we don't have a separate authz object for bundles at the moment -- their authz access is tied to the Fleet level -- I'll change all the "Read-access" actions to `Read`, and all the write-access actions to `Modify`.

(I'll do the same for all the spots mentioned below, too)